node/circuits.js

```diff
@@ -0,0 +1,238 @@
+// healthy and other factors, but the circuit only needs to know about health
+// before forwarding.
+
+function AlwaysShouldRequestHandler() { }
+
+AlwaysShouldRequestHandler.prototype.shouldRequest = function shouldRequest() {
+    return true;
+};
+
+var alwaysShouldRequestHandler = new AlwaysShouldRequestHandler();
+
+//  circuit = circuits
+//      .circuitsByServiceName[serviceName]
+//      .circuitsByCallerName[callerName]
+//      .circuitsByEndpointName[endpointName]
+
+function EndpointCircuits(root) {
+    var self = this;
+    self.root = root;
+    self.circuitsByEndpointName = {};
+}
+
+EndpointCircuits.prototype.getCircuit = function getCircuit(callerName, serviceName, endpointName) {
+    var self = this;
+    var circuit = self.circuitsByEndpointName['$' + endpointName];
+    if (!circuit) {
+        circuit = new Circuit();
+        circuit.callerName = callerName;
+        circuit.serviceName = serviceName;
+        circuit.endpointName = endpointName;
+        circuit.shouldRequestOptions = self.root.shouldRequestOptions;
+        var stateOptions = self.root.stateOptions;
+        circuit.stateOptions = {
+            stateMachine: circuit,
+            nextHandler: stateOptions.nextHandler,
+            timers: stateOptions.timers,
+            random: stateOptions.random,
+            period: stateOptions.period,
+            maxErrorRate: stateOptions.maxErrorRate,
+            minRequests: stateOptions.minRequests,
+            probation: stateOptions.probation
+        };
+        circuit.setState(States.HealthyState);
+        self.circuitsByEndpointName['$' + endpointName] = circuit;
+    }
+    return circuit;
+};
+
+EndpointCircuits.prototype.getCircuitTuples = function () {
+    var self = this;
+    return Object.keys(self.circuitsByEndpointName).map(function (endpointName) {
+        var circuit = self.circuitsByEndpointName[endpointName];
+        return [circuit.callerName, circuit.serviceName, circuit.endpointName];
+    });
+};
+
+function ServiceCircuits(root) {
+    var self = this;
+    self.root = root;
+    self.circuitsByCallerName = {};
+}
+
+ServiceCircuits.prototype.getCircuit = function getCircuit(callerName, serviceName, endpointName) {
+    var self = this;
+    var circuits = self.circuitsByCallerName['$' + callerName];
+    if (!circuits) {
+        circuits = new EndpointCircuits(self.root);
+        self.circuitsByCallerName['$' + callerName] = circuits;
+    }
+    return circuits.getCircuit(callerName, serviceName, endpointName);
+};
+
+ServiceCircuits.prototype.getCircuitTuples = function getCircuitTuples() {
+    var self = this;
+    var circuits = Object.keys(self.circuitsByCallerName).map(function (callerName) {
+        return self.circuitsByCallerName[callerName].getCircuitTuples();
+    });
+    return Array.prototype.concat.apply([], circuits);
+};
+
+function Circuits(options) {
+    var self = this;
+    self.circuitsByServiceName = {};
+    self.stateOptions = {
```

Comment:
Okay, now I've seen it used in two places, should we get a prototype for this? constructor signature `function(stateMachine, options)` and in this case we'd need to set .nextHandler after construction?